### PR TITLE
b2: Add new `cleanup` and `cleanup-hidden` backend commands.

### PR DESCRIPTION
**Pull request by metadaddy**
#### What is the purpose of this change?

This change provides more control over deleting hidden files and unfinished large file uploads with the `b2` remote by adding two new backend commands:

- The `cleanup` backend command removes unfinished large file uploads older than `max-age`, which defaults to 24 hours, e.g. `rclone backend cleanup -o max-age=1h b2:bucket/path/to/object`.
- The `cleanup-hidden` backend command removes any old hidden versions of files, e.g. `rclone backend cleanup-hidden b2:bucket/path/to/dir`.

#### Was the change discussed in an issue or in the forum before?

No. I noticed it was not possible to use the `b2` backend to delete unfinished large file uploads less than 24 hours old, and I decided to rectify the situation, mirroring the `s3` backend commands.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by metadaddy** [comment]
I adapted the existing internal tests for versions and metadata from the `s3` backend to `b2`. In the process, I encountered an inconsistency between the two backends.

Each file has two timestamps:

1. The time it was actually uploaded to the storage service. This is set by the service itself; the API client cannot directly set this value*.
2. The time it was "last modified". The API client can set this value as user-defined metadata.

For S3, (1) is exposed as the `Last-Modified` HTTP header in `GetObject` and `HeadObject` calls, and captured in the `s3` backend as the `lastModified` field of `s3`'s `Object` struct. (2) is exposed as the `mtime` metadata value, via the `X-Amz-Meta-Mtime` HTTP header. The `s3` backend stores this value as `meta["mtime"]` on its `Object` struct and also makes it available via the `ModTime()` method on the same struct.

For B2, (1) is exposed as the `X-B2-Upload-Timestamp` HTTP header in `b2_download_file_by_id` and `b2_download_file_by_name`, and as `uploadTimestamp` in the response to `b2_get_file_info `.  The backend captures the value in the `UploadTimestamp` field of its `api.File` struct. (2) is exposed as the `src_last_modified_millis` metadata value, via the `X-Bz-Info-src_last_modified_millis` HTTP header. The `b2` backend stores this value in the `modTime` field on its `Object` struct, `Info["src_last_modified_millis"]` on the same struct, as well as making it available via the `ModTime()` method.

This doesn't in itself cause any issues, since each backend is internally consistent, but it can make it difficult to understand what's going on when adapting code from one backend to the other! I'm not sure there's any action to take on this, I'm mostly mentioning it here for the benefit of anyone else working on the two backends.

\* B2 actually does allow you to set the upload time when you upload the file, via the [`X-Bz-Custom-Upload-Timestamp` header on `b2_upload_file`](https://www.backblaze.com/apidocs/b2-upload-file#:~:text=X%2DBz%2DCustom%2DUpload%2DTimestamp). You have to contact Backblaze support to have this feature enabled on your account.

File: backend/util.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Package backend contains all the storage system backends.
+package backend
+
+// StringPointerOrNil returns a pointer to the string if non-empty or nil if it is empty
+func StringPointerOrNil(s string) *string {
```
Comment (ncw) [inline backend/util.go:5]: Not sure where this refactor crept in from - it doesn't seem relevant to this PR?

Utilities for use by backends go in `lib/` - maybe `lib/backend` - though not sure a 4 line function merits a whole new package!

File: backend/b2/b2_internal_test.go [hunk 2]
```diff
@@ -170,9 +185,234 @@ func TestParseTimeString(t *testing.T) {
 
 }
 
+// This is adapted from the s3 equivalent.
+func (f *Fs) InternalTestMetadata(t *testing.T) {
```
Comment (ncw) [inline backend/b2/b2_internal_test.go:189]: This looks like it doesn't belong in this commit?

File: backend/test_util.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import (
+	"bytes"
+	"compress/gzip"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// Gz returns a compressed version of its input string
+func Gz(t *testing.T, s string) string {
```
Comment (ncw) [inline backend/test_util.go:12]: Test utilities go in `fstest/fstest.go`

However I think this probably doesn't belong in this commit.

**Comment by metadaddy** [comment]
Thanks, @ncw - I went back and added some testing, based on the existing S3 tests. This required moving some helper functions around. I'll back the new tests out of this PR, incorporate your suggestions, and submit them as a new PR.

**Comment by ncw** [comment]
Give me a ping when done - thank you :-)

File: backend/util.go (hunk 1 above)
Comment (metadaddy) [reply backend/util.go:5]: This was related to the new B2 tests. I've backed it out of this PR.

File: backend/b2/b2.go [hunk 4]
```diff
@@ -2243,9 +2216,80 @@ func (f *Fs) lifecycleCommand(ctx context.Context, name string, arg []string, op
 	return bucket.LifecycleRules, nil
 }
 
-var commandHelp = []fs.CommandHelp{
-	lifecycleHelp,
-}
+var commandHelp = []fs.CommandHelp{{
+	Name:  "lifecycle",
```
[the remark this answers is not in the document]
Comment (metadaddy) [reply backend/b2/b2.go:2220]: Done.

File: backend/b2/b2_internal_test.go (hunk 2 above)
Comment (metadaddy) [reply backend/b2/b2_internal_test.go:189]: I've backed it out of this PR.

File: backend/test_util.go (hunk 3 above)
Comment (metadaddy) [reply backend/test_util.go:12]: This was related to the new B2 tests. I've backed it out of this PR.

**Comment by metadaddy** [comment]
@ncw Ping!

**Comment by metadaddy** [comment]
@ncw I'm pretty sure that test failure is nothing to do with my changes 🤔

**Comment by metadaddy** [comment]
Hi @ncw - just checking in. There's a failing test, but I'm certain it's unrelated to my changes. Can you trigger the action just to make sure? Thanks!

**Review by ncw — APPROVED** [review]
This is looking very good - thank you :-)

I'll merge it now.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.67 

PS yes that test failure is unrelated.
